src/operation/mod.rs

```diff
@@ -434,3 +434,15 @@ pub(crate) enum Retryability {
     Read,
     None,
 }
+
+macro_rules! remove_empty_write_concern {
```

Comment:
Because all of the options structs passed into this macro contain a `write_concern` field, we don't need to pass the actual write concern into this method -- instead, we can access it directly on the options via `$opts.write_concern`. (The compiler enforces that this field exists when it expands the macro calls.) In addition, I think it would be better to treat `$opts` as an `Option<*Options>` here, i.e.:

```
if let Some(ref mut options) = $opts {
    if let Some(ref write_concern) = options.write_concern {
        if write_concern.is_empty() {
            options.write_concern = None;
        }
    }
}
```

and call as such:

```
remove_empty_write_concern!(self.options);
```

This way we can eliminate the need to construct an empty options struct when the value for `options` is `None`.